lib/croma/new1_existence.ex

```diff
@@ -0,0 +1,115 @@
+    end
+
+    @spec get() :: [module]
+    def get() do
+      if exists?() do
+        Agent.get(__MODULE__, &MapSet.to_list/1)
+      else
+        []
+      end
+    end
+
+    @spec stop() :: :ok
+    def stop() do
+      if exists?() do
+        Agent.stop(__MODULE__)
+      else
+        :ok
+      end
+    end
+  end
+
+  def cleanup() do
+    :ok = AssumedModuleStore.stop()
+  end
+
+  # The second argument `compilers` is for testing purpose only.
+  @doc """
+  Returns true if `mod` exports `new/1`, otherwise false.
+
+  Note that `mod` is compiled and loaded if needed.
+
+  When the `:croma` compiler runs after `:elixir` compiler, this function assumes that
+  `mod` has `new/1` even if `mod` is currently unable to be loaded as following cases:
+
+  - `mod` has not been defined yet because `mod` is open or defined later in the same file.
+  - `mod` is involved in cyclic dependency causing compiler deadlock.
+  """
+  @spec has_new1?(module, list(atom)) :: boolean
+  def has_new1?(mod, compilers \\ Mix.Tasks.Compile.compilers()) do
+    cond do
+      ensure_compiled_and_loaded?(mod) ->
+        function_exported?(mod, :new, 1)
+      uses_custom_compiler_after_elixir_compiler?(compilers) ->
+        :ok = AssumedModuleStore.inserted_new(mod)
+        # Assume that `mod` has `new/1`; by this assumption, modules whose
+        # existence of `new/1` depends on existence of `mod.new/1` (e.g.,
+        # modules `use`ing `Croma.SubtypeOfList`) will also have `new/1`.
+        # This is convenient when `mod` and those modules refer to each other.
+        true
+      true ->
+        mod_str = inspect(mod)
+        raise """
+        Cannot determine whether #{mod_str} has new/1 or not. \
+        This might be because #{mod_str} is mutually referred from another module \
+        or #{mod_str} is referred from its child module. \
+        For these cases, try using :croma compiler (you need to put it after :elixir compiler).\
+        """
+    end
+  end
+
+  defp ensure_compiled_and_loaded?(mod) do
+    case Code.ensure_compiled(mod) do
+      {:module, ^mod} ->
+        # Although `Code.ensure_compiled/1` didn't return an error,
+        # `Code.ensure_loaded?/1` can still return false when `mod` is open.
+        Code.ensure_loaded?(mod)
+      {:error, reason} when reason in [:nofile, :unavailable] ->
+        # We expect that `mod` will be compiled in the future, and thus don't raise.
+        # `:nofile` indicates that `mod` is defined later in the same file, and
+        # `:unavailable` indicates that `mod` is involved in cyclic dependency.
+        false
+      {:error, reason} ->
+        raise "Cannot load module #{inspect(mod)} due to reason #{inspect(reason)}"
+    end
+  end
+
+  defp uses_custom_compiler_after_elixir_compiler?(compilers) do
+    with [:elixir | compilers_after_elixir] <- Enum.drop_while(compilers, &(&1 != :elixir)) do
+      :croma in compilers_after_elixir
+    else
+      _ -> false
+    end
```

Comment:
For very rare occasions, users could be using another custom compiler that wraps the standard `:elixir` compiler (such as the one defined in the old version of this PR). Should we support it?

If yes, then we shouldn't require that `:elixir` be present for `:croma` compiler to work. This function would become a simple `in` expression (and thus could be removed).

```suggestion
    :croma in compilers
```

If no, we could `raise` if e.g.
- `:elixir` compiler is not present in `compilers`, or
- `:croma` compiler comes before `:elixir`.

What do you think? In my opinion we should take the "yes" route and avoid introducing such limitation here.

Reply:
I think we should take the "yes" too.
Still, I think it might be helpful to raise if `:croma` compiler comes before `:elixir` compiler because such cases are probably misconfiguration and the developers can fix it by themselves.

I allowed the case where `:elixir` compiler is not present at https://github.com/skirino/croma/pull/12/commits/44aaff8e9b9e330739cd3da44dcf07573275b22a, and add some test cases at https://github.com/skirino/croma/pull/12/commits/a3c268830d78115492801d52a3fbdf335c57d81f.